### PR TITLE
Materials table and friends

### LINKED ISSUE #211: Improve relations between ref_api texturenums and vk_materials
- [x] Pass material reference directly in `vk_render_geometry_t` instead of texture index.
- [x] Separate materials table, not directly related to textures table.
- [ ] ~~Do not use textures name -> index hash table mapping~~ moved to #601

### PR DESCRIPTION
**Pull request by w23**
- [ ] KTX2 #154
  - [x] proof of concept
  - [x] proper alignment for blocks (validation complains)
- [ ] texture leaks: #594 
  - [x] investigate
- [x] `inherit`/`use`
- [x] Fix #211 
  - [x] make materials table
  - [ ] ~~make materials by-name lookup independent of texture table, stop doing dummy textures~~ moved to #601 
- [ ] normalmap glitches, #595
  - [x] investigated, seems to be offline normalization vs bit depth/precision issue; needs offline changes

### CODE REVIEW

**Comment by 0x4E69676874466F78** [comment]
На тему KTX2 замечания:
1. в ab53aae359c7f166ddc22c957fce3f6cfc3657b3 была починена нормаль для BC5_S, но сломана для PNG.
2. нормали грузятся 16bit_BC6H/16bit_BC6H_SF/16bit_16F и даже работают, но работают так же криво как и для PNG.
3. артефакт с красным цветом это из-за `"basecolor_map" "debug_roughness_bc4.ktx2"`, раньше было `"basecolor_map" "debug_roughness.png"`, то есть вместо трёх каналов стал один он естественно прилетает только в красный, остальные оказываются по нулям.

**Comment by w23** [comment]
32-битная сборка ёкнулась, потому что конечно же в 2023 году невозможно универсально printf-напечатать uint64_t

**Comment by a1batross** [comment]
If you want KTX2 support, better support it at imagelib level. Not only everything in the engine goes through imagelib, it's also kinda "shared" between engine and SDK.

It's not that hard, you basically only need to implement the loader for KTX2 file format and store compressed data in the `rgba` field, which then will be compressed at renderer level, see `img_dds.c` for example.

File: ref/vk/vk_textures.c [hunk 1]
```diff
@@ -762,19 +782,293 @@ const byte*	VK_TextureData( unsigned int texnum )
...
 
 	// see if already loaded
-	if(( tex = Common_TextureForName( name )))
+	vk_texture_t *tex = Common_TextureForName( name );
+	if( tex )
 		return (tex - vk_textures);
 
+	{
+		const char *ext = Q_strrchr(name, '.');
+		if (Q_strcmp(ext, ".ktx2") == 0) {
```
Comment (a1batross) [inline ref/vk/vk_textures.c:1061]: I told you about imagelib exactly to avoid this uglyness. :)

Comment (a1batross) [reply ref/vk/vk_textures.c:1061]: Also, to get the file extension, you have `COM_FileExtension` from `crtlib.c`

**Comment by w23** [comment]
> If you want KTX2 support, better support it at imagelib level. Not only everything in the engine goes through imagelib, it's also kinda "shared" between engine and SDK.
> 
> It's not that hard, you basically only need to implement the loader for KTX2 file format and store compressed data in the `rgba` field, which then will be compressed at renderer level, see `img_dds.c` for example.

Yeah, that would definitely need to happen eventually.

I would refrain from doing it in this PR, though, as it's a bit trickier than it sounds, would imply extending imagelib a bit, and would add quite a bit of friction to playing with various texture formats.

This is mostly because KTX2 is highly Vulkan-centric -- it is almost trivial to work with it from Vulkan directly. Imagelib lacks the necessary abstractions to pass enough KTX2 data to preserve this simplicity, unless we're talking about `PF_KTX2` and passing the entire file vanilla with headers in `rgbdata_t.buffer` :zany_face:.

More specifically:
- When using KTX2 internally within ref_vk we can just pass `vkFormat` to the image directly as-is, with its hundreds of possible formats. Not all vulkan formats have valid counterparts of `PF_`. E.g. there are no floating point formats, no `SNORM` vs `UNORM` vs `SRGB` distinction, e.g. there's no `PF_BC7_SRGB` (which we tentatively want) or `PF_BC5_SNORM`, etc.
- Mip layers within KTX2 seem to be laid out differently from DDS, so we either have to pass them separately, or re-shuffle them to accommodate currently expected layout for imagelib users. Re-shuffling would also prevent us from (supposedly possible) single-call upload of the entire mip pyramid as-is.

Making this support now would hurt experimenting with various texture formats -- each new format would require messing with `PF_` definitions, conversions, layouts, sizes, alignments and rebuilding. Now we can just replace the ktx2 file and reload materials (sub-second time).

When we have stabilized the subset of supported formats, we can extract that knowledge to imagelib.

**Comment by w23** [comment]
Also, imagelib is not reentrant, we were contemplating loading textures in multiple threads in background :smiling_face_with_tear:

**Comment by w23** [comment]
- [ ] Кстати, можно явный ремаппинг/свиззл делать для текстур с малым количеством компонент ("красной" шероховатости): https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VkComponentMapping.html

**Comment by a1batross** [comment]
>This is mostly because KTX2 is highly Vulkan-centric 

Doesn't sound like a problem to me. DDS is accelerated-rendering centric, we just don't load it in ref_soft. :)

>unless we're talking about PF_KTX2 and passing the entire file vanilla with headers in rgbdata_t.buffer 

Doesn't sound like a problem to me. If KTX is meant to be Vulkan-centric, so be it.

>Not all vulkan formats have valid counterparts of PF_

It was meant to be extended. My point is that imagelib is always used by the whole engine. The main feature is to be able to omit extension, and give mod developer freedom in file formats, as it probes common file extension, detect file formats and allows in-memory loading on top of it.

>When we have stabilized the subset of supported formats, we can extract that knowledge to imagelib.

I hear you, experiment as much as you want. I'm just telling you that modifying engine for your own needs is fine. For me extending existing subsystems is always more funnier than doing hacky things, it makes development more iterative.

**Pull request merged**